### PR TITLE
master-port 49201

### PR DESCRIPTION
**Pull request by mchugh19**
### What does this PR do?
Port of develop merged #49201 to master

### Tests written?
No - There don't seem to be any tests of recurse_types

### CODE REVIEW

**Review by dwoz — CHANGES_REQUESTED** [review]
@mchugh19 Is it possible to add a unit test for this please?

**Comment by mchugh19** [comment]
> @mchugh19 Is it possible to add a unit test for this please?

As mentioned, there aren't any tests of recurse_types. Are there any suggestions of how to handle this?

**Comment by mchugh19** [comment]
Hey all. Happy to look into tests for this, if there is a recommendation on what they should look like.

**Comment by garethgreenaway** [comment]
@mchugh19 Apologies for the delay in responding.  We did have some existing tests that were testing including two conflicting recurse options, in the unit test for `file.directory`.  I had started off tweaking things to just provide a quick example to pass along but quickly realized that the test in question was rather broken, so in addition to some cleanup I also added a test to test the use of silent as a recurse option.

File: salt/states/file.py [hunk 1]
```diff
@@ -3732,6 +3735,9 @@ def directory(name,
             file_mode = None
             dir_mode = None
 
+        if 'silent' in recurse_set:
+            ret['pchanges'] = 'Changes silenced'
```
Comment (Ch3LL) [inline salt/states/file.py]: a while back we removed pchanges in favor of changes here https://github.com/saltstack/salt/pull/49832

can you change this to changes

File: tests/unit/states/test_file.py [hunk 2]
```diff
@@ -1255,30 +1262,91 @@ def test_directory(self):
...
+                        if salt.utils.platform.is_windows():
+                            mock_perms = MagicMock(return_value=check_perms_ret)
+                        else:
+                            mock_perms = MagicMock(return_value=(check_perms_ret, ''))
+
+                        recurse = ['silent']
+                        ret = {'name': name,
+                               'result': False,
+                               'comment': '',
+                               'pchanges': 'Changes silenced',
```
Comment (Ch3LL) [inline tests/unit/states/test_file.py]: should remove pchanges here as well in favor of changes

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/states/file.py (hunk 1 above)
Comment (mchugh19) [reply salt/states/file.py]: Done! It's a little confusing as master still has pchanges, but hopefully that's all that needs adjusting here.

File: tests/unit/states/test_file.py (hunk 2 above)
Comment (mchugh19) [reply tests/unit/states/test_file.py]: Also done!

**Comment by mchugh19** [comment]
Support output like:
```
----------
          ID: /tmp/bigthing
    Function: file.directory
      Result: True
     Comment: Directory /tmp/bigthing updated
     Started: 17:15:16.822358
    Duration: 416.771 ms
     Changes:   
              ----------
              recursion:
                  Changes silenced
              user:
                  root

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
``` 
where changes is expected to remain a dictionary as later operation will continue to update it.

**Comment by mchugh19** [comment]
re-run pr-pre-commit

**Comment by mchugh19** [comment]
re-run pr-windows2019-py3

**Comment by waynew** [comment]
@dwoz / @Ch3LL tests are passing now 🎉

**Review by dwoz — APPROVED** [review]

**Pull request merged**